### DISCUSSION TITLE
Can I use Kornia morphological operations to build custom loss functions

### DISCUSSION THREAD

**Discussion by MLCOOKER** [kornia/kornia#1626] [Questions]
I know there's already a loss function component in kornia, however, I'd like to build my own loss functions that perform certain morphological operations on predicted and target images. Having briefly looked at building custom losses in pytorch, I gathered I need the functions to be differentiable and built using the pytorch backend (ie not resorting to numpy). 

Can I use the kornia.morphology utilities to, for example, compute the L1 loss between white top hat versions of predicted and target images? Would this violate the computation of gradients and subsequent backpropagation? If so, would this extend to using any utility, for example, the canny edge version of an image?

Thank you in advance for your replies!
B

**Comment by edgarriba** [comment]
@MLCOOKER in principle you can do that. We design and test all the core ops so that you can back-propagate /cc @Manza12

**Reply by Manza12** [reply]
@MLCOOKER All the morphological operations in Kornia are made by composition of PyTorch functions; this means they are differentiable.
I guess that your concern regards the fact that morphological operations are not smooth, but in the computation of gradients you can use this type of operations with some arrangements; in this case, the place where operations are not differentiable is when we take the minimum/maximum, but PyTorch manages that creating a gradient that is zero everywhere but in the place that has given the minimum/maximum.
I hope I have answered to your question. Do not hesitate in asking if you want more details about morphology :)

**Reply by golnooshAbd** [reply]
What about homography estimation and other kornia.geometry operations like kornia.get_perspective_transform? Are they differentiable to use as loss and handle error backpropagation? I use kornia.warp_perspective() as loss,
but _**loss.backward()**_ says that:

> element 0 of tensors does not require grad and does not have a grad_fn

**Reply by Manza12** [reply]
I am not familiar with kornia.get_perspective_transform. Maybe @edgarriba can answer you better.

**Reply by edgarriba** [reply]
in principle it allows backprop -- could you share a snippet showing the error ?
